### PR TITLE
fix(pinning): check before writing root address

### PR DESCRIPTION
**Pull request by notanatol**
### Checklist

- [X] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] I have filled out the description and linked the related issues.

### Description
Closes: #4527

### CODE REVIEW

File: pkg/storer/internal/pinning/pinning.go [hunk 1]
```diff
@@ -243,6 +243,16 @@ func (c *collectionPutter) Put(ctx context.Context, st internal.Storage, writer
 		return nil
 	}
 
+	rootCollection := &pinCollectionItem{Addr: ch.Address()}
+	has, err := st.IndexStore().Has(rootCollection)
+	if err != nil {
+		return fmt.Errorf("pin store: check previous root: %w", err)
+	}
+	if has {
```
Comment (ldeffenb) [inline pkg/storer/internal/pinning/pinning.go:251]: This doesn't solve all of the problems, but only Put's that happen AFTER the initial pin is actually closed.   Consider the following:

1) Two or more simultaneous operations start pinning the exact same content.
2) They all do the individual Puts, and the root collection doesn't yet exist, so they all actually store their collectionChunks.
3) One of the pinners actually executes the Close causing the pinCollectionItem to exist.  That one "wins" and actually completes the pin.
4) The other pinners will now quit Putting their collectionChunks, but their originally stored (raced) records still exist.   And worse, their chunkstore.Put() invocation have increased the chunks' RefCnts.
5) When the other pinners finally invoke Close, they'll correctly not replace the original pinCollectionItem, but all of those incorrect RefCnt increments are still there causing those chunks to never be able to be removed from the chunkstore.

File: pkg/storer/internal/pinning/pinning.go [hunk 2]
```diff
@@ -264,11 +274,19 @@ func (c *collectionPutter) Close(st internal.Storage, writer storage.Writer, roo
...
+	}
+
+	if !has {
+		// Save the root pin reference.
+		c.collection.Addr = root
+		err := writer.Put(c.collection)
+		if err != nil {
+			return fmt.Errorf("pin store: failed updating collection: %w", err)
+		}
 	}
```
Comment (ldeffenb) [inline pkg/storer/internal/pinning/pinning.go:290]: This doesn't solve the problem of residual collectionChunks with their corresponding forever-remaining increments of the chunkstore RefCnt.   Consider:

1) Two or more simultaneous operations start pinning the exact same content.
2) They all do the individual Puts, and the root collection doesn't yet exist, so they all actually store their collectionChunks and invoke chunkstore.Put incrementing RefCnt on those chunks.
3) One of the pinners actually executes the Close causing the pinCollectionItem to exist.  That one "wins" and actually completes the pin.
4) When the other pinners finally invoke Close, they'll correctly not replace the original pinCollectionItem, but all of those incorrect RefCnt increments are still there causing those chunks to never be able to be removed from the chunkstore.

**Review by ldeffenb — COMMENTED** [review]
I corrected my original review by moving that comment to Close() where it belongs, but added a new concern with the Has() check that was added to Put().

File: pkg/storer/internal/pinning/pinning.go (hunk 1 above)
Comment (notanatol) [reply pkg/storer/internal/pinning/pinning.go]: it's been removed now

File: pkg/storer/internal/pinning/pinning.go (hunk 2 above)
Comment (notanatol) [reply pkg/storer/internal/pinning/pinning.go:284]: Basically returning an error on line :276 will make `cleanupOnErrWriter` wrapper object to perform a cleanup (by calling the `Cleanup` method) whenever the error is detected
https://github.com/ethersphere/bee/blob/bf56d82e1aff5d5e9fde395044a410e9b8baf03f/pkg/api/api.go#L870-L874
notice the `onErr` pointing to the `.Cleanup` method
https://github.com/ethersphere/bee/blob/bf56d82e1aff5d5e9fde395044a410e9b8baf03f/pkg/api/bzz.go#L106-L110
which in turn will decrement the ref count(s) to the correct value. See the test below.

Comment (ldeffenb) [reply pkg/storer/internal/pinning/pinning.go:284]: Thank you for that description.   I should have followed the error path back out from Close() myself.  I'm setting up a testnet environment to verify this on a full OSM tileset load whenever it is merged.

**Review by ldeffenb — APPROVED** [review]
Been tracking this one as it developed and the current version looks like it should solve the pinning refCnt issue that it intends to solve.   Looking forward to testing it once it merges.

**Review by acha-bill — APPROVED** [review]

**Pull request merged**